yb-voyager/src/migassessment/gather.go

```diff
@@ -0,0 +1,581 @@
+
+	// Channel for progress updates
+	progressChan := make(chan NodeProgress, totalNodes*10) // Buffer for multiple updates per node
+
+	// Channel to signal completion of progress display goroutine
+	displayDone := make(chan struct{})
+
+	// Start progress display goroutine
+	go func() {
+		defer close(displayDone)
+		for progress := range progressChan {
+			tracker.update(progress)
+		}
+	}()
+
+	// WaitGroup for parallel collection
+	var wg sync.WaitGroup
+
+	// Channel for collection results
+	resultChan := make(chan collectionResult, totalNodes)
+
+	// Start parallel collection for all nodes
+	for _, node := range nodes {
+		wg.Add(1)
+		go func(n collectionNode) {
+			defer wg.Done()
+
+			// Run collection with buffered output
+			err := runGatherAssessmentMetadataScriptBuffered(
+				scriptPath,
+				[]string{fmt.Sprintf("PGPASSWORD=%s", source.Password), "PGCONNECT_TIMEOUT=10"},
+				n.nodeName,
+				n.displayName,
+				progressChan,
+				assessmentMetadataDir,
+				n.connectionUri,
+				source.Schema,
+				assessmentMetadataDir,
+				fmt.Sprintf("%t", pgssEnabled),
+				fmt.Sprintf("%d", iopsInterval),
+				"true",     // --yes (doesn't matter since skip_checks=true)
+				n.nodeName, // source_node_name
+				"true",     // skip_checks - guardrails already validated
+			)
+
+			// Send result
+			resultChan <- collectionResult{
+				displayName: n.displayName,
+				isPrimary:   n.isPrimary,
+				err:         err,
+			}
+		}(node)
+	}
+
+	// Goroutine to close channels after all collections complete
+	go func() {
+		wg.Wait()
+		close(progressChan)
+		close(resultChan)
+	}()
+
+	// Wait for display goroutine to finish (it closes when progressChan is closed)
+	<-displayDone
+
+	// Process results
+	var failedReplicasList []string
+	var successfulReplicaCount int
+	var primaryFailed bool
+	var primaryErr error
+
+	for result := range resultChan {
+		if result.err == nil {
+			log.Infof("Successfully collected metadata from %s", result.displayName)
+			if !result.isPrimary {
+				successfulReplicaCount++
+			}
+		} else {
+			if result.isPrimary {
+				primaryFailed = true
+				primaryErr = result.err
+			} else {
+				log.Warnf("Failed to collect metadata from replica %s: %v", result.displayName, result.err)
+				failedReplicasList = append(failedReplicasList, result.displayName)
+			}
+		}
+	}
+
+	// If primary failed, return error immediately
+	if primaryFailed {
+		return nil, fmt.Errorf("metadata collection failed on primary database (critical): %w", primaryErr)
+	}
+
+	// Print summary
+	if !primaryFailed && len(failedReplicasList) > 0 {
+		fmt.Println() // Blank line before warnings
+		color.Yellow("WARNING: Metadata collection failed on %d replica(s): %v", len(failedReplicasList), failedReplicasList)
+		utils.PrintAndLogfInfo("Continuing assessment with data from primary + %d successful replica(s)", successfulReplicaCount)
+		utils.PrintAndLogfWarning("Note: Sizing and metrics will reflect only the nodes that succeeded.")
+	}
```

Comment:
why just warn and move on?
why not fail?

Reply:
I think it's simpler to just fail. Why complicate it? If it fails, the user gets to know what happend. If they want to exclude that replica, they can then just explicitly specify the ones to include via a config.